62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+	static std::array<double, 4> solveQuarticRoots(double a, double b, double c, double d, double e, double t_start, double t_end)
+	{
+		std::array<double, 4> t = { -1.0, -1.0, -1.0, -1.0 }; // only two candidates in range, ever
+
+		const double quadCoeffMag = std::max(std::abs(d), std::abs(e));
+		const double cubCoeffMag = std::max(std::abs(c), quadCoeffMag);
+		if (std::abs(a) > std::max(std::abs(b), cubCoeffMag) * QUARTIC_THRESHHOLD)
+		{
+			auto res = equations::Quartic<double>::construct(a, b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 4);
+		}
+		else if (abs(b) > quadCoeffMag)
+		{
+			auto res = equations::Cubic<double>::construct(b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 3);
+		}
+		else
+		{
+			auto res = equations::Quadratic<double>::construct(c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 2);
+		}
+
+		if (!(t[0] != t[1]))
+			t[0] = t[0] != t_start ? t_start : t_end;
+
+		//// TODO: check that this clamp works with t[i] as NaN
+		//for (auto i = 0; i < 2; i++)
+		//	t[i] = nbl::core::clamp(t[i], t_start, t_end);
+		//
+		//// fix up a fuckup where both roots are NaN or were on the same side of the valid integral
+		//
+		//// TODO: do some Halley or Householder method steps on t
+		////while ()
+		////{
+		////}
+
+		// neither t still not in range, your beziers don't intersect
+		return t;
+	}
+
+	// (only works for monotonic curves)
+	static float64_t2 getCurveRoot(double p0, double p1, double p2)
+	{
+		// https://people.csail.mit.edu/bkph/articles/Quadratics.pdf
+		double a = p0 - 2.0 * p1 + p2;
+		double b = 2.0 * (p1 - p0);
+		double c = p0;
+
+		// Solve linear equation
+		if (a == 0.0) return float64_t2(-c / b);
+
+		double det = sqrt(b * b - 4 * a * c);
+		if (b >= 0.0) {
+			return float64_t2(
+				(- b - det) / (2 * a),
+				(2 * c) / (-b - det)
+			);
+		}
+		else
+		{
+			return float64_t2(
+				(2 * c) / (-b + det),
+				(-b + det) / (2 * a)
+			);
+		}
+		//double det = b * b - 4 * a * c;
+		//double rcp = 0.5 / a;
+		//
+		//double detSqrt = sqrt(det) * rcp;
+		//double brcp = b * rcp;
+		//
+		//// Solve linear equation
+		//if (isinf(brcp)) return float64_t2(-c / b);
+		//
+		//return float64_t2(-detSqrt, detSqrt) - brcp;
+	}
+};
+
+Hatch::QuadraticBezier Hatch::Segment::splitCurveRange(QuadraticBezier curve, double left, double right) const
+{
+    return curve.splitCurveTakeLower(right).splitCurveTakeUpper(left / right);
+}
+
+Hatch::QuadraticBezier Hatch::Segment::getSplitCurve() const
+{
+    return splitCurveRange(*originalBezier, t_start, t_end);
+}
+
+bool Hatch::Segment::isStraightLineConstantMajor() const
+{
+	auto major = (uint)SelectedMajorAxis;
+	return originalBezier->p[0][major] == originalBezier->p[1][major] &&
+		originalBezier->p[0][major] == originalBezier->p[2][major];
+}
+
+std::array<double, 2> Hatch::Segment::intersect(const Segment& other) const
```

Comment:
I think two bugs in this function:

1. you don't handle the case where the other bezier is line segment but the current one isn't
2. you don't take into account the segments are between t_start and t_end and 0.0 and 1.0, so you shouldn't return what intersectOrtho returns directly but check against t_start and t_end

the reason these slipped from your hand was probably due to your test cases not having many non line bezier intersections. 
make sure to add tests where you have curves intersection a lot and include a case like the image here
![image](https://i.ibb.co/WtdSQJK/image.png)

Reply:
the reason for the specific case above is it's messing a bit with the pre requisited of the optimized sweep line algorithm and is a good test case